src/System.Private.ServiceModel/tests/Common/Infrastructure/src/BridgeClientCertificateManager.cs

```diff
@@ -0,0 +1,394 @@
+
+                    string base64Cert = string.Empty;
+                    response = BridgeClient.MakeResourceGetRequest(MachineCertificateResourceName, requestParams);
+
+                    string certificateAsBase64;
+                    if (response.TryGetValue(CertificateKeyResourceString, out certificateAsBase64))
+                    {
+                        certificateToInstall = new X509Certificate2(Convert.FromBase64String(certificateAsBase64));
+                    }
+                    else
+                    {
+                        StringBuilder sb = new StringBuilder();
+                        foreach (var pair in response)
+                        {
+                            sb.AppendFormat("{0}  {1} : {2}", Environment.NewLine, pair.Key, pair.Value);
+                        }
+
+                        throw new Exception(
+                            string.Format("Error retrieving {0} certificate from Bridge. Expected '{1}' key in response. Response contents:{2}{3}",
+                                s_LocalFqdn,
+                                CertificateKeyResourceString,
+                                Environment.NewLine,
+                                sb.ToString()));
+                    }
+                }
+            }
+
+            InstallCertificateToMyStore(certificateToInstall);
+
+            // We also need to install the root cert if we install a local cert
+            InstallRootCertificateFromBridge();
+        }
+
+        // Uninstalls all certificates and SSL port associations
+        // added by this process.  If 'force' is true, removes all,
+        // whether this process created them or not.
+        public static void UninstallAllCertificates(bool force)
+        {
+            lock (s_certificateLock)
+            {
+                UninstallAllMyCertificates(force);
+                UninstallAllRootCertificates(force);
+            }
+        }
+
+        // Returns the certificate matching the given thumbprint from the given store.
+        // Returns null if not found.
+        private static X509Certificate2 CertificateFromThumbprint(X509Store store, string thumbprint)
+        {
+            X509Certificate2Collection foundCertificates = store.Certificates.Find(X509FindType.FindByThumbprint, thumbprint, validOnly: true);
+            return foundCertificates.Count == 0 ? null : foundCertificates[0];
+        }
+
+        // Adds the given certificate to the given store unless it is
+        // already present.  Returns 'true' if the certificate was added.
+        private static bool AddToStoreIfNeeded(StoreName storeName,
+                                               StoreLocation storeLocation,
+                                               X509Certificate2 certificate)
+        {
+            X509Store store = null;
+            X509Certificate2 existingCert = null;
+            try
+            {
+                store = new X509Store(storeName, storeLocation);
```

Comment:
Should this code be inside a lock?  Can it be re-entered, and if so are there any issues with opening a Store ReadWrite multiple times concurrently?